lib/runtime/wasmer/imports.go

```diff
@@ -325,62 +325,187 @@ func ext_misc_runtime_version_version_1(context unsafe.Pointer, z C.int64_t) C.i
 }
 
 //export ext_default_child_storage_read_version_1
-func ext_default_child_storage_read_version_1(context unsafe.Pointer, a C.int64_t, b C.int64_t, c C.int64_t, d C.int32_t) C.int64_t {
+func ext_default_child_storage_read_version_1(context unsafe.Pointer, childStorageKey C.int64_t, key C.int64_t, valueOut C.int64_t, offset C.int32_t) C.int64_t {
 	logger.Trace("[ext_default_child_storage_read_version_1] executing...")
-	logger.Warn("[ext_default_child_storage_read_version_1] unimplemented")
-	return 0
+
+	instanceContext := wasm.IntoInstanceContext(context)
+	storage := instanceContext.Data().(*runtime.Context).Storage
+	memory := instanceContext.Memory().Data()
+
+	value, err := storage.GetChildStorage(asMemorySlice(instanceContext, childStorageKey), asMemorySlice(instanceContext, key))
+	if err != nil {
+		logger.Error("[ext_default_child_storage_read_version_1] failed to get child storage", "error", err)
+		return 0
+	}
+
+	valueBuf, valueLen := int64ToPointerAndSize(int64(valueOut))
+	copy(memory[valueBuf:valueBuf+valueLen], value[offset:])
+
+	size := uint32(len(value[offset:]))
+	sizeBuf := make([]byte, 4)
+	binary.LittleEndian.PutUint32(sizeBuf, size)
+
+	sizeSpan, err := toWasmMemoryOptional(instanceContext, sizeBuf)
+	if err != nil {
+		logger.Error("[ext_default_child_storage_read_version_1] failed to allocate", "error", err)
+		panic(err)
```

Comment:
instead of panicking can you return 0? same for all other functions

Reply:
We are panicking at other places where we use `toWasmMemoryOptional`. Should I change them as well?

Reply:
yes you can change them